### DISCUSSION TITLE
Allow configuring non-manual sort order for the o2m style interfaces

### DISCUSSION THREAD

**Discussion by paulboudewijn** [directus/directus#12771] [Feature Requests]
When no sort field is configured for a relation, the PK is used as the default sort field instead of using the field(s) used in the display template. This results in a seemingly random list of items in the m2m interface.

I have a PR ready.

### To Reproduce

1. Create a 'book' collection with the 'title' and 'subtitle' fields
2. Create an 'author' collection
3. Create a m2m relation between 'author' and 'book', without sortfield and interface displaytemplate
4. Note that the m2m interface list 'books' on the 'author' collection is sorted by the PK field
5. Add displaytemplate `{{book_id.title}} - {{book_id.subtitle}}` to the interface
6. Note that the m2m interface list is still sorted in the PK field, although title and subtitle alphabetically could suggest otherwise
7. Add a sort field  `sort` to the relation
8. Note that the interface list is sorted on this column (as expected)

**Comment by oreilles** [comment]
This isn't a bug: the primary key is used so that when no sort field is defined, the items appear in creation order.
Using the template instead is an interesting idea, but the discussions are a better fit for this kind of request.
I will close this issue, but feel free to submit it again in the discussion section.

**Comment by paulboudewijn** [comment]
> This isn't a bug: the primary key is used so that when no sort field is defined, the items appear in creation order.
> Using the template instead is an interesting idea, but the discussions are a better fit for this kind of request.

Well, the following line tells me it was intended to use the (first) displaytemplate field when the sortField wasn't defined. 
https://github.com/directus/directus/blob/a920041c5b12eab72b5daeb71eaea11129cec8f8/app/src/interfaces/list-m2m/use-sort.ts#L18

But this line prevents that:
https://github.com/directus/directus/blob/a920041c5b12eab72b5daeb71eaea11129cec8f8/app/src/interfaces/list-m2m/use-sort.ts#L22

Looks like a bug to me, but it's now discussion #9929

**Comment by rijkvanzanten** [comment]
@Oreilles thoughts?

**Comment by oreilles** [comment]
> Well, the following line tells me it was intended to use the (first) displaytemplate field when the sortField wasn't defined.
> But this line prevents that:

That seems right... Was using the first field template as a default sort field actually intended ? Using the primary key to use insertion order instead makes more sense to me. Also, if no manual sort is set, shouldn't the sort be done in the api ? That would explain line 22..

**Comment by paulboudewijn** [comment]
> Using the primary key to use insertion order instead makes more sense to me.

What benefit does a user have to see m2m items sorted on a identity value instead of alphabetically on the name of this item?

> Also, if no manual sort is set, shouldn't the sort be done in the api ?

Yes absolutely, that's also how I initially thought it would work.

**Comment by rijkvanzanten** [comment]
> What benefit does a user have to see m2m items sorted on a identity value instead of alphabetically on the name of this item?

None at all, but it does make sense as the default for "sort field = null". That being said, I think it'd be a great optimization if we could default the sortField _option_ to be the first field in the display template, so it's explicit, but still have the logic default to whatever the default is in the API when the sortField option is set to `null`

**Comment by paulboudewijn** [comment]
> we could default the sortField option to be the first field in the display template, so it's explicit

By this you mean if SortField is empty and a display template is set, SortField gets automatically filled with the first field used in the display template? I was under the impression SortField was used to store the sort order after manually resorting the items via drag-n-drop. Feels tricky to me to use the same field to select a sorting field in a (related) table from which you don't want the value to be overwritten after drag-n-drop.

> whatever the default is in the API 

This default would then be the sort field as configured in a preset of the junction table used in the m2m relation?

**Comment by rijkvanzanten** [comment]
> By this you mean if SortField is empty and a display template is set, SortField gets automatically filled with the first field used in the display template? I was under the impression SortField was used to store the sort order after manually resorting the items via drag-n-drop. Feels tricky to me to use the same field to select a sorting field in a (related) table from which you don't want the value to be overwritten after drag-n-drop.

Ah yeah my bad. I was confusing this with an option to set the display sort-value for when you don't have a manual drag-and-drop sort field configured under the relationship settings, which is currently not a thing yet but something I'd envision could be part of the sprint to add https://github.com/directus/directus/pull/9135

**Comment by natanparmigiano** [comment]
Hey Everyone, 

Was looking on the sorting issue today and realized something that **might** have flown under the radars regarding the default sorting behaviour being the ID, that being collections that use UUIDs.

That is the case for my application, and because of that, when new items are created in the relation they seem to be randomly ordered (since uuids have no intrinsic ordering), creating a quite strange behaviour for the end user.

Would'nt it make sense to add a new separate "sort by" field, and disable it if the collection is made sortable by means of drag'n'drop?

Im using the o2m interface for a revisions and approval workflow where the creation order is the expected ordering, but since i have to use UUIDs (i expose the IDs as API endpoints in a custom extension) being able to simply allow ordering by the date_created field without the risk of users mangling the collection using the (uselessly enabled in my case) drag and drop functionality would be really nice!


EDIT: 

I just now realized that using the default sort option in the interface does not allow reverse ordering (newest to oldest) for example.

That seems quite strange since the API itself allows for arbitrary sorting of the results

**Reply by natanparmigiano** [reply]
Also, since [#9135](https://github.com/directus/directus/pull/9135) has already been merged, and it seems that the non manual sort-value idea has not been implemented, would it be a good idea to revive this issue to track the potential implementation of that item as a feature?

I would'nt mind taking a shot at it myself, im not yet so familiar with the codebase, but i think that might be a good spot to start :) (as long as no one more familiar with the involved moving parts is not already working on it).

**Comment by u12206050** [comment]
Not sure how this conversation died, but could we add a sort field option on the interface option?

**Reply by rijkvanzanten** [reply]
What would happen when you'd configure both this interface-sort and the relationship-sort? Would the interface-sort be ignored?

**Reply by u12206050** [reply]
Correct, would think that is the most logical or?

**Reply by rijkvanzanten** [reply]
I think so too, might be a bit confusing though if there's two options called `sort` in two different panes of the same field setup. What would be a better name? "Default Sort" with a field note that says "Sort used if manual sort isn't enabled for this field" ?

**Reply by u12206050** [reply]
Could we hide the "Default Sort" field when a sorting field exists on the related collection?

**Reply by u12206050** [reply]
Although I guess one could always add a sorting field afterwards as well. But if I am not mistaken, the manual sort field is on the collection level, but the sort field we are talking about now is on the relationship level, ie. you could choose a different field to sort on if you want on a different relationship even though it is the same collection.

**Reply by u12206050** [reply]
So think on the interface it will be fine with calling it "sort", but as hint we could say: ""Sort used if manual sort isn't enabled for this collection"

**Reply by u12206050** [reply]
Just found out I could edit the "raw value" on the collection level to select a field that is not an integer to sort by. Seems to work as I need :)

**Comment by rijkvanzanten** [comment]
This feature request has been accepted and is queued to be implemented! You can follow along with the progress here: https://github.com/directus/directus/issues/18076
